### PR TITLE
Revamp Proxy Contract

### PR DESCRIPTION
**Pull request by scooprinder**
# Description
This removes any references to faith from the Vault Proxy contract, moves to use solidity custom errors (see #483) and lastly introduces a simple mechanism to allow users to withdraw early from a vault, forfeiting any potential yield. 

# Checklist
- [x] Code follows the style guide
- [x] I have performed a self-review of my own code
- [x] New and existing tests pass locally
- [x] This PR is targeting the correct branch

### CODE REVIEW

**Comment by scooprinder** [comment]
@princetonbishop tagged you as you raised the custom error issue and I just wanted to make sure I was using them as intended (looks like in 0.8.4, custom errors are not compatible with `require` - https://blog.soliditylang.org/2021/04/21/custom-errors/ )

File: protocol/contracts/core/VaultProxy.sol [hunk 1]
```diff
@@ -113,23 +71,49 @@ contract VaultProxy is Ownable {
...
     }
+
+    /**
+     * Allows users to withdraw their Temple from a vault before the exit window, forfeits any yield
+     */
+    function exitVaultEarly(uint256 amount, Vault vaultErc) external {
+        SafeERC20.safeTransferFrom(IERC20(vaultErc), msg.sender, address(this), amount);
+        uint256 shareAmount = vaultErc.toSharesAmount(amount);
+        uint256 templeBal = vaultErc.toTokenAmount(shareAmount);
+        SafeERC20.safeTransfer(temple, msg.sender, templeBal);
```
Comment (medariox) [inline protocol/contracts/core/VaultProxy.sol:98]: Would love to have an event here if possible, with something like:
`emit EarlyWithdraw(msg.sender, templeBal, vaultErc)`
`event EarlyWithdraw(address account, uint256 amount, address vault)`

**Comment by scooprinder** [comment]
Scope change; want to funnel exits through LBP. Want to keep vault proxy clean of this so will create a separate contract.

**Comment by princetonbishop** [comment]
> @princetonbishop tagged you as you raised the custom error issue and I just wanted to make sure I was using them as intended (looks like in 0.8.4, custom errors are not compatible with `require` - https://blog.soliditylang.org/2021/04/21/custom-errors/ )

@scooprinder yes. recommended to use `revert CustomError(...params)`. The stax repo is a good place to check in case you want conformity with stax.

File: protocol/contracts/core/VaultProxy.sol [hunk 2]
```diff
@@ -6,75 +6,37 @@ import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
...
-    /** @notice Tokens / Contracted required for the proxy contract  */
+    // Tokens / Contracted required for the proxy contract 
     IERC20 public immutable ogTemple;
     IERC20 public immutable temple;
     TempleStaking public immutable templeStaking;
-    Faith public immutable faith;
-    bool public faithClaimEnabled;
+
+    // Errors
+    error NoTempleReceivedWhenUnstaking();
```
Comment (princetonbishop) [inline protocol/contracts/core/VaultProxy.sol:23]: maybe some params here might help? eg. `NoTempleReceivedWhenUnstaking(balBefore, balAfter)` or something. could help with some testing

File: protocol/contracts/core/VaultProxy.sol [hunk 3]
```diff
@@ -113,23 +77,56 @@ contract VaultProxy is Ownable {
...
+
+    /**
+     * @notice Allows owner to redeem any vault shares from users that have exited early
+     */
+    function redeemVaultTokenToTemple(Vault vaultErc) public onlyOwner {
+        if (!vaultErc.canExit()) {
+            revert CanNotExitVault();
+        }
+        uint256 balance = vaultErc.balanceOf(address(this));
+        vaultErc.withdraw(balance);
```
Comment (princetonbishop) [inline protocol/contracts/core/VaultProxy.sol:122]: if (balance > 0)

File: protocol/contracts/core/VaultProxy.sol [hunk 4]
```diff
@@ -97,7 +59,9 @@ contract VaultProxy is Ownable {
         uint256 templeBeforeBalance = temple.balanceOf(address(this));
         templeStaking.unstake(_amountOGT);
         uint256 templeAfterBalance = temple.balanceOf(address(this));
-        require(templeAfterBalance > templeBeforeBalance, "Vault Proxy: no Temple received when unstaking");
+        if (templeAfterBalance < templeBeforeBalance) {
+            revert NoTempleReceivedWhenUnstaking();
+        }
 
         return templeAfterBalance - templeBeforeBalance;
```
Comment (princetonbishop) [inline protocol/contracts/core/VaultProxy.sol:66]: nit: unchecked

**Pull request closed without merging**